### PR TITLE
Proposed fix for #9634

### LINKED ISSUE #9634: `formatResults` on belongsToMany doesn't apply to results
This is a (multiple allowed):
* [x] bug
* [ ] enhancement
* [ ] feature-discussion (RFC)

* CakePHP Version: 3.3.6

### What you did

Added a callback to a belongsToMany query's `formatResults`.

### What happened

The callback never fires. I modified the test case that supposedly tests this feature to actually test it. The test case didn't actually test that the callback fires. Here's the failing test commit: https://github.com/jeremyharris/cakephp/commit/02f5af9ca7717a900e3011d05688ac40f0b3f6ca

You can place a `die` statement in that callback and see that it will never get called.

### Expected Behavior

`formatResults` to work on belongsToMany.

I'd submit a PR but I spent a few hours trying to figure out why and couldn't. Perhaps someone could point me in the right direction?

### PR DESCRIPTION
**Pull request by jeremyharris**
This fixes #9634, however I have to admit I don't know if this will have any potential side effects I'm not considering. In the `testFormatResultsBelongsToMany` test I added a formatter to the originating table to make sure it gets called still and doesn't get the belongsToMany formatter applied to it, just in case.

### CODE REVIEW

File: src/ORM/Association.php [hunk 1]
```diff
@@ -815,6 +815,11 @@ protected function _formatAssociationResults($query, $surrogate, $options)
         $formatters = $surrogate->formatResults();
 
         if (!$formatters || empty($options['propertyPath'])) {
+            if ($formatters) {
```
Comment (markstory) [inline src/ORM/Association.php]: These nested conditionals are a bit odd, as `if($formatters)` is inside `if(!$formatters)`. Perhaps the new `if($formatters)` case can be hoisted above this conditional?

**Review by markstory — COMMENTED** [review]
Tests Look good to me.

File: src/ORM/Association.php (hunk 1 above)
Comment (jeremyharris) [inline src/ORM/Association.php]: Yeah it's a bit weird. The alternatives seem odd too, like if empty property path return, if no formatters return else apply formatters and return. Or if formatters and empty property path apply formatters then return else return.

Seems like there'd be an else in there... Or a whole extra outside double conditional.

Comment (josegonzalez) [inline src/ORM/Association.php]: This is definitely more readable:

```php
if (!$formatters) {
	return;
}

if (empty($options['propertyPath'])) {
    foreach ($formatters as $formatter) {
        $query->formatResults($formatter);
    }
    return;
}
```

Comment (jeremyharris) [inline src/ORM/Association.php]: There it is. Drunk jeremy is drunk.

File: tests/TestCase/ORM/QueryTest.php [hunk 2]
```diff
@@ -1328,13 +1344,17 @@ public function testFormatResultsBelongsToMany()
...
+        $this->assertNull($first->joinBeforeFind);
+
         $this->assertCount(2, $first->tags);
         $expected = [
             'id' => 1,
             'name' => 'tag1',
             '_joinData' => ['article_id' => 1, 'tag_id' => 1],
             'description' => 'A big description',
             'created' => new Time('2016-01-01 00:00'),
+            'joinBeforeFind' => true,
```
Comment (lorenzo) [inline tests/TestCase/ORM/QueryTest.php]: Why is this here and not in `_joinData`? I don't understand how after being applied to `articles_tags` you expect the results to be here

File: tests/TestCase/ORM/QueryTest.php [hunk 3]
```diff
@@ -1310,11 +1310,27 @@ public function testFormatResultsBelongsToMany()
...
                 });
-            }, 'Model.beforeFind');
+            });
+
+        $table
+            ->eventManager()
+            ->on('Model.beforeFind', function ($event, $query) {
+                $query->formatResults(function ($results) {
+                    foreach ($results as $result) {
+                        $result->beforeFind = true;
```
Comment (lorenzo) [inline tests/TestCase/ORM/QueryTest.php]: What happened to this property? I don't see it in the tests.

File: tests/TestCase/ORM/QueryTest.php (hunk 2 above)
Comment (jeremyharris) [inline tests/TestCase/ORM/QueryTest.php]: Ok. I originally had it in _joinData (a much easier fix) but I wrote the original test case incorrectly and followed it. Oy what a waste. At any rate, I'll rebase.

I've got to check our app code to make sure the issue I originally discovered is even related to this one now :)

**Comment by jeremyharris** [comment]
Rebased and removed the commits that formatted the wrong property. `_joinData` is now formatted as it's the join table.

**Comment by markstory** [comment]
Looks good to me.

**Comment by lorenzo** [comment]
Thanks!

**Pull request merged**